generator/json-generator.lua

```diff
@@ -1,135 +1,156 @@
+local json = require 'dkjson'
+
 -- ------------------------------------------------
 -- Constants
 -- ------------------------------------------------
 
-local APOSTROPHE =  '"'
-local COMMA      =  ','
-local LINE_BREAK = '\n'
-local TAB        = '  '
-
-local PARAM  = '${%d:%s (%s)}'
-
 local OUTPUT_FILE = 'love-completions.json'
-
 local WIKI_URL = 'https://love2d.org/wiki/'
 
+local LOVE_MODULE_STRING = 'love.'
+
+-- JSON Output control
+local DEBUG = false
+local KEY_ORDER = {
+    'type',
+    'description',
+    'fields',
+    'args',
+    'returnTypes',
+    'link'
+}
+
 -- ------------------------------------------------
 -- Local Functions
 -- ------------------------------------------------
 
-local function generateArguments( arguments, start )
-    local params = ''
-    start = start or 0
+---
+-- Generates a sequence containing subtables of all the arguments
+--
+local function createArguments( args )
+    local arguments = {}
+    for i, v in ipairs( args ) do
+        arguments[i] = {}
+        arguments[i].name = v.name
+
+        -- Use [foo] notation as display name for optional arguments.
+        if v.default then
+            arguments[i].displayName = string.format( '[%s]', v.name )
+        end
+    end
+    return arguments
+end
+
+local function getType( type )
+    if type == 'string'
+    or type == 'number'
+    or type == 'table'
+    or type == 'boolean'
+    or type == 'function' then
+        return { type = type }
+    end
+    return { type = 'ref', name = type }
+end
 
-    if arguments then
-        for i, args in ipairs( arguments ) do
-            params = params .. string.format( PARAM, start + i, args.name, args.type )
+local function createReturnTypes( returns )
+    local returnTypes = {}
+    for i, v in ipairs( returns ) do
+        returnTypes[i] = getType( v.type )
+    end
+    return returnTypes
+end
 
-            -- Add a separator unless we are dealing with the last argument of the function.
-            if i ~= #arguments then
-                params = params .. ', '
-            end
-        end
+-- TODO handle functions with neither args nor return variables.
+local function createVariant( vdef )
+    local variant = {}
+    if vdef.arguments then
+        variant.args = createArguments( vdef.arguments )
     end
```

Comment:
Oh that's great! What are you trying? Using `__index` referencing the SuperType, and the table part as the Types's methods? That should work at least for the methods repetitions